app/src/main/java/ch/hikemate/app/model/facilities/FacilitiesViewModel.kt

```diff
@@ -3,17 +3,130 @@ package ch.hikemate.app.model.facilities
+   * performs several checks to determine which facilities should be displayed:
+   * 1. Verifies if the zoom level is sufficient for showing facilities
+   * 2. Checks if the center of the view is close enough to the hike route
+   * 3. Limits the number of displayed facilities based on the current zoom level
+   *
+   * The function uses lazy evaluation through sequences to efficiently process facilities, stopping
+   * as soon as the maximum number of facilities for the current zoom level is reached.
+   *
+   * @param facilities List of all available facilities to filter
+   * @param bounds The current visible bounds of the map view
+   * @param zoomLevel The current zoom level of the map
+   * @param hikeRoute The detailed hike route being displayed
+   * @param onSuccess Callback invoked with the filtered list of facilities when processing succeeds
+   * @param onNoFacilitiesForState Callback invoked when no facilities should be displayed for the
+   *   current state
+   */
+  fun filterFacilitiesForDisplay(
+      facilities: List<Facility>,
+      bounds: BoundingBox,
+      zoomLevel: Double,
+      hikeRoute: DetailedHike,
+      onSuccess: (List<Facility>) -> Unit,
+      onNoFacilitiesForState: () -> Unit
+  ) {
+    viewModelScope.launch(dispatcher) {
+      // 1. Early returns for invalid states
+      if (!isValidZoomAndBounds(zoomLevel, bounds, hikeRoute)) {
+        onNoFacilitiesForState()
+        return@launch
+      }
+
+      // 2. Calculate max facilities once
+      val maxFacilities = getMaxFacilitiesForZoom(zoomLevel)
+      if (maxFacilities == 0) {
+        onNoFacilitiesForState()
+        return@launch
+      }
+
+      // 3. Use sequence for lazy evaluation and take() for early termination
+      val filteredFacilities =
+          facilities
+              .asSequence()
+              .filter { facility ->
+                bounds.contains(GeoPoint(facility.coordinates.lat, facility.coordinates.lon))
+              }
+              .take(maxFacilities)
+              .toList()
+
+      onSuccess(filteredFacilities)
+    }
+  }
+
+  /**
+   * Validates whether facilities should be displayed for the current zoom level and map bounds.
+   * This function performs two main checks:
+   * 1. Verifies if the zoom level meets the minimum threshold for displaying facilities
+   * 2. Ensures the center point of the view is within the maximum allowed distance from the route
+   *
+   * @param zoomLevel Current zoom level of the map
+   * @param bounds Current visible bounds of the map
+   * @param hikeRoute The detailed hike route being displayed
+   * @return true if facilities should be displayed for the current state, false otherwise
+   */
+  private fun isValidZoomAndBounds(
+      zoomLevel: Double,
+      bounds: BoundingBox,
+      hikeRoute: DetailedHike
+  ): Boolean {
+    // This might seem redundant but the reason this is used is so that if the zoom level
```

Comment:
This check is redundant for the ZoomLevel condition. Here's why:
- This function checks whether the zoom is high enough >13
- As you can see in line 65:
 ```kotlin 
val maxFacilities = getMaxFacilitiesForZoom(zoomLevel)
```
This is a "double check" per say because we check if the map contains any facilities whatsoever for that specific `zoomLevel`. That's why it might be deemed "redundant" for zoom levels outside that map.